### ISSUE TITLE
Fault on running Tempesta FW w/ configured, but not loaded Frang

### ISSUE THREAD

**Issue by krizhanovsky** [tempesta-tech/tempesta#888]
Config file, Frang limits are configured:
```
listen 192.168.100.4:80;
server 127.0.0.1:8080; #apache
cache 0;

frang_limits {
	http_ct_vals "text/plain" "text/html";
	client_header_timeout 1;
	client_body_timeout 1;
	request_rate 20;
	request_burst 15;
	http_host_required false;
}
```

However, Tempesta FW is started w/o Frang module:
```
# ./scripts/tempesta.sh --restart
```
which leads to the system crash:
```
[ 3716.047307] [tempesta] ERROR: don't know how to handle: 'frang_limits'
[ 3716.048160] [tempesta] ERROR: configuration parsing error:
[ 3716.048160]    6: frang_limits {
[ 3716.048160]       ^^^^^^^^^^^^^^
[ 3716.051545] general protection fault: 0000 [#1] SMP DEBUG_PAGEALLOC
[ 3716.052146] Modules linked in: tfw_sched_ratio(O) tfw_sched_http(O) tfw_sched_hash(O) tempesta_fw(O) tempesta_db(O) tempesta_tls(O) binfmt_misc bochs_drm ttm drm_kms_helper crct10dif_pclmul ghash_clmulni_intel drm aesni_intel aes_x86_64 lrw gf128mul glue_helper ablk_helper cryptd ppdev sg fb_sys_fops syscopyarea sysfillrect sysimgblt serio_raw parport_pc parport pcspkr button ip_tables x_tables autofs4 ext4 crc16 jbd2 fscrypto mbcache sr_mod sd_mod cdrom ata_generic ata_piix libata psmouse scsi_mod e1000 i2c_piix4
[ 3716.052146] CPU: 2 PID: 1044 Comm: sysctl Tainted: G           O    4.9.35-kdump #131
[ 3716.052146] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.10.2-2.fc27 04/01/2014
[ 3716.052146] task: ffff97eaba9ca6c0 task.stack: ffffbcdb408b4000
[ 3716.052146] RIP: 0010:[<ffffffffc07bc29e>]  [<ffffffffc07bc29e>] tfw_sg_release+0xe/0x80 [tempesta_fw]
[ 3716.052146] RSP: 0018:ffffbcdb408b7c88  EFLAGS: 00010202
[ 3716.052146] RAX: ffff97eaba9ca6c0 RBX: ffff97eaa0891508 RCX: 0000000000000006
[ 3716.052146] RDX: 0000000000000006 RSI: 0000000000000001 RDI: 6b6b6b6b6b6b6b6b
[ 3716.052146] RBP: ffffbcdb408b7ca8 R08: d84156c5635688c0 R09: ffff97eb20891e60
[ 3716.052146] R10: 0000000000000000 R11: 0000000000000000 R12: 6b6b6b6b6b6b6b6b
[ 3716.052146] R13: 0000000000000001 R14: ffffffffc07d5e00 R15: 00000000ffffffea
[ 3716.052146] FS:  00007fdc4de17880(0000) GS:ffff97eabf900000(0000) knlGS:0000000000000000
[ 3716.052146] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 3716.052146] CR2: 00007fdc4d5f0b10 CR3: 0000000060893000 CR4: 00000000003406e0
[ 3716.052146] Stack:
[ 3716.052146]  ffff97eaa0891508 0000000000000001 0000000000000001 ffffffffc07d5e00
[ 3716.052146]  ffffbcdb408b7cc8 ffffffffc07c0fd1 ffffffffc07d68b0 0000000000000001
[ 3716.052146]  ffffbcdb408b7cf8 ffffffffc07c1088 ffffffffc07d6300 ffffffffc07d6300
[ 3716.052146] Call Trace:
[ 3716.052146]  [<ffffffffc07c0fd1>] tfw_cfgop_cleanup_srv_cfg+0x91/0xb0 [tempesta_fw]
[ 3716.052146]  [<ffffffffc07c1088>] tfw_cfgop_cleanup_srv_cfgs+0x98/0xc0 [tempesta_fw]
[ 3716.052146]  [<ffffffffc07c10c3>] tfw_cfgop_cleanup_srv_groups+0x13/0x20 [tempesta_fw]
[ 3716.052146]  [<ffffffffc07995f3>] spec_cleanup+0x53/0x70 [tempesta_fw]
[ 3716.052146]  [<ffffffffc079a82c>] tfw_cfg_cleanup+0x1c/0x30 [tempesta_fw]
[ 3716.052146]  [<ffffffffc07ba01e>] tfw_ctlfn_state_io+0x11e/0x480 [tempesta_fw]
[ 3716.052146]  [<ffffffffc07b9f41>] ? tfw_ctlfn_state_io+0x41/0x480 [tempesta_fw]
[ 3716.052146]  [<ffffffff81e6af04>] ? ns_capable_common+0x34/0x90
[ 3716.052146]  [<ffffffff8204e025>] proc_sys_call_handler+0xe5/0x100
[ 3716.052146]  [<ffffffff8204e04f>] proc_sys_write+0xf/0x20
[ 3716.052146]  [<ffffffff81fca322>] __vfs_write+0x32/0x160
[ 3716.052146]  [<ffffffff81ecb683>] ? rcu_read_lock_sched_held+0x53/0x60
[ 3716.052146]  [<ffffffff81ecb98a>] ? rcu_sync_lockdep_assert+0x2a/0x50
[ 3716.052146]  [<ffffffff81fce96d>] ? __sb_start_write+0xad/0x1a0
[ 3716.052146]  [<ffffffff81fcb728>] ? vfs_write+0x178/0x190
[ 3716.052146]  [<ffffffff81fcb660>] vfs_write+0xb0/0x190
[ 3716.052146]  [<ffffffff81fcca83>] SyS_write+0x53/0xc0
[ 3716.052146]  [<ffffffff823dd72e>] entry_SYSCALL_64_fastpath+0x1c/0xb1
[ 3716.052146] Code: 7d c0 89 45 ec e8 33 13 c2 c1 8b 45 ec 48 83 c4 10 5b 41 5c 5d c3 31 c0 eb e1 0f 1f 00 55 48 89 e5 41 56 41 55 41 54 53 49 89 fc <48> 8b 47 68 48 85 c0 74 0b 48 8b 40 20 48 85 c0 74 02 ff d0 4d
[ 3716.052146] RIP  [<ffffffffc07bc29e>] tfw_sg_release+0xe/0x80 [tempesta_fw]
[ 3716.052146]  RSP <ffffbcdb408b7c88>
[ 3716.114966] ---[ end trace f94c5a73aec8fa85 ]---
```

The fix should be to correctly cleanup and release the data on the wrong start and/or adjust `tempesta.sh` script that it parses the config file and decides automatically to add `-f` option (in this case we can remove the option).

Personally, I stepped several times on incoordinated configuration and the command line option for Frang.

**Comment by ikoveshnikov** [comment]
Can't reproduce the issue using `./scripts/tempesta.sh --restart`. All i get is just configuration load error:
```
[  155.476990] [tempesta] New configuration is cleaned.
[  157.842372] [tempesta] Warning: Trying to stop an inactive system
[  157.848531] [tempesta] Prepearing for the configuration processing.
[  157.850471] [tempesta] ERROR: don't know how to handle: 'frang_limits'
[  157.852117] [tempesta] ERROR: configuration parsing error:
[  157.852117]    5: frang_limits {
[  157.852117]       ^^^^^^^^^^^^^^
[  157.857480] [tempesta] Configuration parsing has failed. Clean up...
[  157.859318] [tempesta] New configuration is cleaned.
```
But that is normal situation. I've tried more and caught a crash during restart without unloading TempestaFW modules. Instead `./scripts/tempesta.sh --restart` command the next one should be used: `sysctl -w net.tempesta.state=stop; sysctl -w net.tempesta.state=start`. After _several_ (not just first) resets I get a crush. But the crush is little bit different:
```
[  130.956303] [tempesta] Warning: Trying to stop an inactive system
[  132.285013] [tempesta] Warning: Trying to stop an inactive system
[  134.576607] [tempesta] Prepearing for the configuration processing.
[  134.581857] [tempesta] ERROR: don't know how to handle: 'frang_limits'
[  134.584079] [tempesta] ERROR: configuration parsing error:
[  134.584079]    5: frang_limits {
[  134.584079]       ^^^^^^^^^^^^^^
[  134.587764] [tempesta] Configuration parsing has failed. Clean up...
[  134.589786] [tempesta] New configuration is cleaned.
[  136.009999] [tempesta] Warning: Trying to stop an inactive system
[  137.137276] [tempesta] Prepearing for the configuration processing.
[  137.141010] ------------[ cut here ]------------
[  137.144033] kernel BUG at /home/user/tempesta/tempesta/tempesta_fw/classifier.c:88!
[  137.148914] invalid opcode: 0000 [#1] SMP
[  137.150058] Modules linked in: tfw_sched_ratio(O) tfw_sched_http(O) tfw_sched_hash(O) tempesta_fw(O) tempesta_db(O) tempesta_tls(O) crct10dif_pclmul crc32_pclmul snd_hda_codec_generic binfmt_misc iTCO_wdt iTCO_vendor_support ghash_clmulni_intel ppdev snd_hda_intel snd_hda_codec snd_hda_core snd_hwdep snd_pcm snd_timer qxl evdev serio_raw pcspkr snd soundcore ttm lpc_ich mfd_core virtio_console sg virtio_balloon parport_pc parport drm_kms_helper drm shpchp button ip_tables x_tables autofs4 ext4 crc16 jbd2 crc32c_generic fscrypto ecb mbcache sr_mod cdrom virtio_blk virtio_net crc32c_intel aesni_intel ahci libahci aes_x86_64 glue_helper lrw gf128mul ablk_helper cryptd psmouse libata i2c_i801 i2c_smbus ehci_pci uhci_hcd ehci_hcd usbcore usb_common sym53c8xx scsi_transport_spi scsi_mod virtio_pci virtio_ring virtio
[  137.177027] CPU: 1 PID: 962 Comm: sysctl Tainted: G           O    4.9.0-tempesta-amd64 #1 Debian 4.9.35-tfw4-1
[  137.180270] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.11.0-20171110_100015-anatol 04/01/2014
[  137.183148] task: ffff9919b777ab80 task.stack: ffffa7efc07ec000
[  137.184828] RIP: 0010:[<ffffffffc076ddc2>]  [<ffffffffc076ddc2>] tfw_classifier_add_inport+0x22/0x30 [tempesta_fw]
[  137.187609] RSP: 0018:ffffa7efc07efa60  EFLAGS: 00010246
[  137.189239] RAX: 0000000000000007 RBX: ffff9919b4e05e40 RCX: 0000000000000000
[  137.191416] RDX: ffffffffc07ab270 RSI: ffffffffc07ab270 RDI: 0000000000005000
[  137.193641] RBP: ffffa7efc07efb18 R08: ffffffffc07ab270 R09: ffff9919b4e05e40
[  137.195780] R10: 0000000000000000 R11: 0000000000000004 R12: 0000000000000000
[  137.197848] R13: 0000000000000000 R14: ffffa7efc07efbe8 R15: ffffffffc07ab1c0
[  137.200028] FS:  00007fd625b12880(0000) GS:ffff9919bfd00000(0000) knlGS:0000000000000000
[  137.202472] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  137.204225] CR2: 00007fd62520cd00 CR3: 0000000075032000 CR4: 00000000003406e0
[  137.206047] Stack:
[  137.206840]  ffffffffc07914d6 ffffffffc07ab250 ffffa7efc07efb18 0000000000000000
[  137.209033]  ffffffffc07915cd f281ca087cd347d3 ffffa7efc07efac8 ffff9919b559b340
[  137.211230]  f281ca087cd347d3 ffffffff9246aa9a ffffa7efc07efbe8 ffff9919b559b340
[  137.213543] Call Trace:
[  137.214310]  [<ffffffffc07914d6>] ? tfw_listen_sock_add.part.9+0x76/0xb0 [tempesta_fw]
[  137.216714]  [<ffffffffc07915cd>] ? tfw_listen_sock_add+0xbd/0xe0 [tempesta_fw]
[  137.218959]  [<ffffffff9246aa9a>] ? get_acl+0x1a/0xf0
[  137.220501]  [<ffffffffc0791752>] ? tfw_cfgop_listen+0x162/0x190 [tempesta_fw]
[  137.222696]  [<ffffffffc076cec4>] ? parse_cfg_entry+0x154/0x400 [tempesta_fw]
[  137.225385]  [<ffffffffc076d113>] ? parse_cfg_entry+0x3a3/0x400 [tempesta_fw]
[  137.227788]  [<ffffffff92412989>] ? terminate_walk+0x89/0xf0
[  137.229509]  [<ffffffff924177cb>] ? path_openat+0x76b/0x1440
[  137.231264]  [<ffffffffc076cc59>] ? spec_handle_entry+0x79/0x90 [tempesta_fw]
[  137.233429]  [<ffffffffc076d78d>] ? tfw_cfg_parse_mods+0x18d/0x210 [tempesta_fw]
[  137.235204]  [<ffffffffc076d043>] ? parse_cfg_entry+0x2d3/0x400 [tempesta_fw]
[  137.236521]  [<ffffffffc076db8f>] ? tfw_cfg_parse+0x2f/0x50 [tempesta_fw]
[  137.238686]  [<ffffffffc078c5a5>] ? tfw_ctlfn_state_io+0x105/0x530 [tempesta_fw]
[  137.241365]  [<ffffffff9280b77e>] ? mutex_lock+0xe/0x30
[  137.243007]  [<ffffffffc078c4e4>] ? tfw_ctlfn_state_io+0x44/0x530 [tempesta_fw]
[  137.245245]  [<ffffffff92482b43>] ? proc_sys_call_handler+0xe3/0x100
[  137.247251]  [<ffffffff92408820>] ? vfs_write+0xb0/0x190
[  137.248663]  [<ffffffff92409c12>] ? SyS_write+0x52/0xc0
[  137.250210]  [<ffffffff9280e1fb>] ? system_call_fast_compare_end+0xc/0x9b
[  137.251858] Code: 83 c4 08 c3 0f 1f 44 00 00 0f 1f 44 00 00 8b 05 a5 e0 03 00 83 f8 07 74 12 8d 50 01 66 89 bc 00 40 be 7a c0 89 15 8f e0 03 00 c3 <0f> 0b 66 90 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 48 83 
```

I've looked more and it seems that the more simple configuration is affected:
```
listen 192.168.100.4:80;
```
And again, I can't reproduce the crash with `--restart` command, only using pure sysctl interface and only after several attempts. 

It seems to me that the issue itself consist of two parts: feature request for loading `frang` on demand (if it's listed in configuration file, done in #920 ) and duplicate of #893 point 3. But I couldn't reproduce the original crash and I can miss something.

**Comment by krizhanovsky** [comment]
Yeah, maybe so. Also we have #918 which is also not 100% reproducable.

**Comment by ikoveshnikov** [comment]
Frang on-demand start is fixed in d5cabf39e704ca1dbda073f41ea01b0b5bd7c073
Issue in classifier to be closed in #893
